### PR TITLE
bpo-44603: Exit the interpreter if the user types "exit"

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue44603

### CODE REVIEW

File: Python/pythonrun.c [hunk 1]
```diff
@@ -267,6 +285,13 @@ PyRun_InteractiveOneObjectEx(FILE *fp, PyObject *filename,
         }
         return -1;
     }
+
+    if (PyRun_IsInteractiveExitCommand(mod)) {
+        PyErr_SetString(PyExc_SystemExit, "bye!");
```
Comment (pganssle) [inline Python/pythonrun.c]: Does this need a string? Typing `exit` doesn't print anything, it just closes the REPL.

File: Python/pythonrun.c [hunk 2]
```diff
@@ -188,6 +188,24 @@ PyRun_InteractiveLoopFlags(FILE *fp, const char *filename, PyCompilerFlags *flag
...
+        return 0;
+    }
+    expr_ty expr = statement->v.Expr.value;
+    if (expr->kind != Name_kind) {
+        return 0;
+    }
+    if (expr->v.Name.ctx != Load) {
+        return 0;
+    }
+    return PyUnicode_CompareWithASCIIString(expr->v.Name.id, "exit") == 0;
```
Comment (Fidget-Spinner) [inline Python/pythonrun.c]: The ``quit()`` people feel a little lonely here 😉 .

File: Python/pythonrun.c (hunk 1 above)
Comment (pablogsal) [reply Python/pythonrun.c]: This is just to show that ir works :)

File: Python/pythonrun.c (hunk 2 above)
Comment (pablogsal) [reply Python/pythonrun.c]: If we decide to go with this, I will promise to finish the PR 😉  (There are some other details missing)

Comment (Fidget-Spinner) [reply Python/pythonrun.c]: Right. In any case, thanks for taking this issue up! I don't think there's such a precedent for special repl-only parsing, so I'm impressed to see a working implementation.

**Comment by gpshead** [comment]
LGTM, I suggest reopening this (button clicked).

From a testing perspective, we just want to make sure it doesn't _interfere_ with with what uber interactive console add-on's like IPython already implement that do this.

**Comment by gpshead** [comment]
From a PR standpoint I think the thing we need was covered in your bug comment:

"if there's a single AST node that is a name and the name is "exit" or "quit", the REPL inspects locals and globals to see if the object referred to is a Quitter, and if so it exits"

We should be able to add unittest coverage for this in `test_repl.py` including equivalent test coverage that such exit/quit treatment is not present outside the repl.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
For the purpose of the review, I will be neutral on the proposal.    But I do notice that the special behavior is limited to entering exactly 'quit\n' or  'exit\n'.  This is stricter than current ^Z behavior.
```
>>> a^Z
  File "<stdin>", line 1
    a→
     ^
SyntaxError: invalid syntax
>>> ^Zb

f:\dev\3x>
```
Exiting instead of also raising SyntaxError may be a bug.

Changes:
1. The reprs of both functions should be changed to eliminate the ()s.
2. The code class imitating the REPL should imitate this also.  If no syntax error, run same tests on result as ast.parse.  But do in a separate function, as here, so subclass can easily disable with def is_exit(code): return False.

File: Python/pythonrun.c [hunk 3]
```diff
@@ -188,6 +188,25 @@ PyRun_InteractiveLoopFlags(FILE *fp, const char *filename, PyCompilerFlags *flag
...
+    expr_ty expr = statement->v.Expr.value;
+    if (expr->kind != Name_kind) {
+        return 0;
+    }
+    if (expr->v.Name.ctx != Load) {
+        return 0;
+    }
+    return PyUnicode_CompareWithASCIIString(expr->v.Name.id, "exit") == 0 ||
+           PyUnicode_CompareWithASCIIString(expr->v.Name.id, "quit") == 0;
+}
```
Comment (terryjreedy) [inline Python/pythonrun.c:209]: Tests of the object associated with the names is needed to not turn them into semi-keywords, as in the the following.
```
>>> exit = 3
>>> exit

f:\dev\3x>
```

**Comment by Privat33r-dev** [comment]
@pablogsal this PR looks great, any chance you will continue working on it? :)

**Comment by pablogsal** [comment]
> @pablogsal this PR looks great, any chance you will continue working on it? :)

There is still no agreement on the solution. Furthermore we are working on a new REPL for CPython that won't need this hack so probably that's the better rule if we decide to go that way

**Comment by pablogsal** [comment]
Superseded by the work in #111201

**Pull request closed without merging**